prov/verbs/src/fi_verbs.h

```diff
@@ -956,7 +969,34 @@ vrb_send_iov_flags(struct vrb_ep *ep, struct ibv_send_wr *wr,
 	if (flags & FI_FENCE)
 		wr->send_flags |= IBV_SEND_FENCE;
 
-	return vrb_post_send(ep, wr, flags);
+	if (wr->send_flags & IBV_SEND_INLINE && hmem != FI_HMEM_SYSTEM) {
+		bounce_buf = malloc(len);
+		if (!bounce_buf)
+			return -FI_ENOMEM;
+
+		ret = ofi_copy_from_hmem_iov(bounce_buf, len, iov, hmem, count,
+					     0);
+		if (ret != len) {
+			if (ret >= 0) {
+				VERBS_WARN(FI_LOG_EP_DATA,
+					   "Unexpected short copy");
+				ret = -FI_EIO;
+			}
+
+			goto out;
+		}
+
+		wr->sg_list[0] = vrb_init_sge(bounce_buf, len, NULL);
+		wr->num_sge = 1;
+	}
+
+	ret = vrb_post_send(ep, wr, flags);
+
+out:
+	if (bounce_buf)
+		free(bounce_buf);
+
+	return ret;
```

Comment:
We need to create separate send calls to use when hmem is enabled.  In most cases, hmem will not be used, and we do not want to add the hmem overhead to all applications.

You can use the performance hooking provider to measure the software overhead the series adds.  Install the perf hook and run the pingpong test.

Reply:
I'll look into this.

Reply:
Here is the performance diff I see:

Without this patchset:
```
libfabric:165371:verbs:core:ofi_perfset_log():138<trace>        PERF: CPU cycles
libfabric:165371:verbs:core:ofi_perfset_log():139<trace>        Name                Avg       Events
libfabric:165371:verbs:core:ofi_perfset_log():148<trace>        perf_recv           1620.29   42
libfabric:165371:verbs:core:ofi_perfset_log():148<trace>        perf_send           1139.65   20
libfabric:165371:verbs:core:ofi_perfset_log():148<trace>        perf_sendmsg        3866      1
libfabric:165371:verbs:core:ofi_perfset_log():148<trace>        perf_inject         4270.9    20
libfabric:165371:verbs:core:ofi_perfset_log():148<trace>        perf_cq_read        307.571   1802

libfabric:165431:verbs:core:ofi_perfset_log():138<trace>        PERF: CPU instr
libfabric:165431:verbs:core:ofi_perfset_log():139<trace>        Name                Avg       Events
libfabric:165431:verbs:core:ofi_perfset_log():148<trace>        perf_recv           1495.79   42
libfabric:165431:verbs:core:ofi_perfset_log():148<trace>        perf_send           878       20
libfabric:165431:verbs:core:ofi_perfset_log():148<trace>        perf_sendmsg        989       1
libfabric:165431:verbs:core:ofi_perfset_log():148<trace>        perf_inject         2648.45   20
libfabric:165431:verbs:core:ofi_perfset_log():148<trace>        perf_cq_read        419.561   1758
```

With this patchset:
```
libfabric:165761:verbs:core:ofi_perfset_log():138<trace>        PERF: CPU cycles
libfabric:165761:verbs:core:ofi_perfset_log():139<trace>        Name                Avg       Events
libfabric:165761:verbs:core:ofi_perfset_log():148<trace>        perf_recv           2013.5    42
libfabric:165761:verbs:core:ofi_perfset_log():148<trace>        perf_send           1196.5    20
libfabric:165761:verbs:core:ofi_perfset_log():148<trace>        perf_sendmsg        3680      1
libfabric:165761:verbs:core:ofi_perfset_log():148<trace>        perf_inject         4513.95   20
libfabric:165761:verbs:core:ofi_perfset_log():148<trace>        perf_cq_read        302.162   1970

libfabric:165829:verbs:core:ofi_perfset_log():138<trace>        PERF: CPU instr
libfabric:165829:verbs:core:ofi_perfset_log():139<trace>        Name                Avg       Events
libfabric:165829:verbs:core:ofi_perfset_log():148<trace>        perf_recv           1591.33   42
libfabric:165829:verbs:core:ofi_perfset_log():148<trace>        perf_send           889       20
libfabric:165829:verbs:core:ofi_perfset_log():148<trace>        perf_sendmsg        1027      1
libfabric:165829:verbs:core:ofi_perfset_log():148<trace>        perf_inject         2657.1    20
libfabric:165829:verbs:core:ofi_perfset_log():148<trace>        perf_cq_read        419.873   1734
```

Still want the send operation split up?

Reply:
If this can be done in a clean way, without a blind copy-paste-modify, then yes.  That's a significant bump on the recv path, which is interesting.

Reply:
>If this can be done in a clean way, without a blind copy-paste-modify, then yes.

This is my concern. 

> That's a significant bump on the recv path, which is interesting.

I need to rerun on these nodes to figure out why recv has the large bump. When I ran on another pair of nodes with debug disabled, here is what I saw:

Without this patchset:
```
libfabric:141685:verbs:core:ofi_perfset_log():139<trace>        Name                Avg       Events
libfabric:141685:verbs:core:ofi_perfset_log():148<trace>        perf_recv           605.357   42
libfabric:141685:verbs:core:ofi_perfset_log():148<trace>        perf_send           499       20
libfabric:141685:verbs:core:ofi_perfset_log():148<trace>        perf_sendmsg        564       1
libfabric:141685:verbs:core:ofi_perfset_log():148<trace>        perf_inject         1150.65   20
libfabric:141685:verbs:core:ofi_perfset_log():148<trace>        perf_cq_read        226.53    1512
```

With this patchset:
```
libfabric:141629:verbs:core:ofi_perfset_log():139<trace>        Name                Avg       Events
libfabric:141629:verbs:core:ofi_perfset_log():148<trace>        perf_recv           609.357   42
libfabric:141629:verbs:core:ofi_perfset_log():148<trace>        perf_send           509       20
libfabric:141629:verbs:core:ofi_perfset_log():148<trace>        perf_sendmsg        589       1
libfabric:141629:verbs:core:ofi_perfset_log():148<trace>        perf_inject         1150.7    20
libfabric:141629:verbs:core:ofi_perfset_log():148<trace>        perf_cq_read        225.46    1683
```

Reply:
I was able to get some minor improvements.

```
libfabric:149726:verbs:core:ofi_perfset_log():139<trace>        Name                Avg       Events
libfabric:149726:verbs:core:ofi_perfset_log():148<trace>        perf_recv           608.357   42
libfabric:149726:verbs:core:ofi_perfset_log():148<trace>        perf_send           508.05    20
libfabric:149726:verbs:core:ofi_perfset_log():148<trace>        perf_sendmsg        577       1
libfabric:149726:verbs:core:ofi_perfset_log():148<trace>        perf_inject         1150.65   20
libfabric:149726:verbs:core:ofi_perfset_log():148<trace>        perf_cq_read        226.868   1465
```

Seems like the next be improvement would be the copy-paste-modify which I don't think we want to do.

Regarding the previous large instruction increase with recv, I am no longer seeing a ~100 plus instruction count. It is ~10-12.

Reply:
I didn't realize the first numbers were from a debug build.  I would ignore debug build analysis.  I agree that we definitely don't want copy-paste-modify, especially since I recently went through the process of consolidating the transfer paths.

Reply:
My mistake for running performance with debug. So here are the current numbers without debug.

Without patchset:
```
libfabric:141685:verbs:core:ofi_perfset_log():139<trace>        Name                Avg       Events
libfabric:141685:verbs:core:ofi_perfset_log():148<trace>        perf_recv           605.357   42
libfabric:141685:verbs:core:ofi_perfset_log():148<trace>        perf_send           499       20
libfabric:141685:verbs:core:ofi_perfset_log():148<trace>        perf_sendmsg        564       1
libfabric:141685:verbs:core:ofi_perfset_log():148<trace>        perf_inject         1150.65   20
libfabric:141685:verbs:core:ofi_perfset_log():148<trace>        perf_cq_read        226.53    1512
```

With patchset:
```
libfabric:149726:verbs:core:ofi_perfset_log():139<trace>        Name                Avg       Events
libfabric:149726:verbs:core:ofi_perfset_log():148<trace>        perf_recv           608.357   42
libfabric:149726:verbs:core:ofi_perfset_log():148<trace>        perf_send           508.05    20
libfabric:149726:verbs:core:ofi_perfset_log():148<trace>        perf_sendmsg        577       1
libfabric:149726:verbs:core:ofi_perfset_log():148<trace>        perf_inject         1150.65   20
libfabric:149726:verbs:core:ofi_perfset_log():148<trace>        perf_cq_read        226.868   1465
```

Is this increase acceptable?

Reply:
Performance numbers tend to be noisy, and highly dependent on the state of the system and network at the time of the job. Are these single measurements, or average measurements from several runs?

Reply:
I have run fi_pingpong multiple times with the hooking provider enabled. Instruction counts returned are close (if the the same) across all runs.